### ISSUE TITLE
[REVIEW]: Enhanced software and platform for the Town Energy Balance (TEB) model

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#2008]
**Submitting author:** @dmey (<a href="http://orcid.org/0000-0002-7071-7547">D. Meyer</a>)
**Repository:** <a href="https://github.com/teb-model/teb" target ="_blank">https://github.com/teb-model/teb</a>
**Version:** 4.0.0
**Editor:** @mjsottile
**Reviewers:** @ethan-nelson, @jayten
**Archive:** <a href="https://doi.org/10.5281/zenodo.3887081" target="_blank">10.5281/zenodo.3887081</a>

## Status

[![status](https://joss.theoj.org/papers/8dc656aecb7a72b80da3bc695342fdb3/status.svg)](https://joss.theoj.org/papers/8dc656aecb7a72b80da3bc695342fdb3)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/8dc656aecb7a72b80da3bc695342fdb3"><img src="https://joss.theoj.org/papers/8dc656aecb7a72b80da3bc695342fdb3/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/8dc656aecb7a72b80da3bc695342fdb3/status.svg)](https://joss.theoj.org/papers/8dc656aecb7a72b80da3bc695342fdb3)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/teb-model/teb" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@jayten & @ethan-nelson, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @mjsottile know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/teb-model/teb">repository url</a>?
- [ ] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@dmey) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/teb-model/teb">repository url</a>?
- [ ] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@dmey) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @rasmussn, @ethan-nelson it looks like you're currently assigned to review this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
PDF failed to compile for issue #2008 with the following error: 

 Can't find any papers to compile :-(

**Comment by dmey** [comment]
@whedon generate pdf from branch joss-paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss-paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02008/joss.02008/10.21105.joss.02008.pdf)

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1023/A:1002463829265 is OK
- 10.5194/gmd-5-1377-2012 is OK
- 10.5194/gmd-11-1929-2018 is OK
- 10.1002/joc.3415 is OK
- 10.5194/gmd-5-433-2012 is OK
- 10.5194/gmd-10-2801-2017 is OK
- 10.1175/WAF-D-11-00064.1 is OK
- 10.3389/fenvs.2014.00014 is OK
- 10.21105/joss.01137 is OK
- 10.5065/1dfh-6p97 is OK
- 10.1016/j.aeaoa.2019.100042 is OK
- 10.1007/s00703-008-0320-9 is OK
- 10.21105/joss.01468 is OK
- 10.1007/BF01025401 is OK
- 10.1007/s007030070003 is OK
- 10.1175/2009MWR2750.1 is OK
- 10.1007/s10546-009-9414-2 is OK
- 10.1175/1520-0493(1998)126<1373:TOCMGE>2.0.CO;2 is OK
- 10.1023/A:1016509614936 is OK
- 10.1007/s10546-006-9091-3 is OK
- 10.1175/2009JAMC2153.1 is OK
- 10.5281/zenodo.3562310 is OK

MISSING DOIs

- https://doi.org/10.1175/1520-0450(2004)043<0312:mtsebo>2.0.co;2 may be missing for title: Modeling the Surface Energy Balance of the Core of an Old Mediterranean City: Marseille
- https://doi.org/10.1007/s00585-997-0090-6 may be missing for title: The Meso-NH Atmospheric Simulation System. Part I: Adiabatic Formulation and Control Simulations
- https://doi.org/10.1175/1520-0450(2002)041<1011:eotteb>2.0.co;2 may be missing for title: Evaluation of the Town Energy Balance (TEB) Scheme with Direct Measurements from Dry Districts in Two Cities

INVALID DOIs

- None
```

**Comment by dmey** [comment]
@whedon generate pdf from branch joss-paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss-paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02008/joss.02008/10.21105.joss.02008.pdf)

**Comment by dmey** [comment]
@mjsottile, @rasmussn, @ethan-nelson I have added the three missing DOIs. Will keep the summary paper on a separate branch (`joss-paper`) in case of any changes and will only merge it into `master` once you are all happy with it.

**Comment by dmey** [comment]
Hi @mjsottile, @rasmussn, @ethan-nelson I was wondering if you had a chance to look at this? Thanks

**Comment by mjsottile** [comment]
:wave: @rasmussn, @ethan-nelson Please let me know if you are having any trouble with the items on the checklist at the top of this thread for the review.

**Comment by ethan-nelson** [comment]
@mjsottile @dmey working through it--should have things completed by tomorrow US time!

**Comment by ethan-nelson** [comment]
This is my first review with JOSS, so thanks for this opportunity. Code looks to be in good shape. Tests run fine and physical values in the raw output of testing are reasonable. Modules are documented on the corresponding website, model scientific documentation is available, and namelist parameters are defined. The paper covers the important points of the software and highlights where it aligns with existing packages.

Responding to a few “unchecked” points in the checklist that are minor: 

## General checks

* License: The OSI list only indicates that CECILL-2.1 is open source compatible. It looks like there are some differences between CECILL-1, and I'm not sure if the OSI designation is backwards-compatible, so I will defer to @mjsottile for that.

## Functionality

* Installation: (CentOS 7.7, cmake 3.12.2, gcc 4.8.5) I ran into issues during compilation:

```
/home/enelson/review/teb/external/minimal-dx/src/shared/minimal_dx_fan.f90:99.64:

      error stop 'Fan Mode must either be 1 for on or 0 for off'
                                                                1
Error: ERROR STOP statement not allowed in PURE procedure at (1)
gmake[2]: *** [CMakeFiles/teb.dir/external/minimal-dx/src/shared/minimal_dx_fan.f90.o] Error 1
gmake[1]: *** [CMakeFiles/teb.dir/all] Error 2
gmake: *** [all] Error 2
```

After removing the error message, the model compiled just fine. This is a fairly old gcc version though so it may not warrant a change.

## Documentation 

* Automated tests: this is a corner case, but it is useful if the user could specify a custom cmake/gcc/git binary location for tests, either in the python call or near the top of one of the python scripts. My cmake is not in my PATH variable so I had to manually change the location in the python code. Adding this would allow for faster iteration over different versions of gcc, etc. for testing compatibility.

* Community guidelines: Might I suggest a CONTRIBUTING.md file or indication in the README to complement the journal text? That could help direct folks who stumble upon the repo first as opposed to the paper.

## Other feedback:

* A file in `docs/` with brief documentation detailing the necessary `input/` forcing files would be useful for tinkering from scratch. I ended up copying files from the example to play around with, but it would be good to have a list all the file variables and units--similar to the namelist docs.

* Something else to consider: what are your thoughts on writing out a simple metadata file in the `output/` directory of a given run? Perhaps this onus should fall on the user to document their runs, but I could see potentially pulling in some metadata from the model build time (e.g. `git rev-parse`) that you could copy to `output/META.txt` or the like.

**Comment by dmey** [comment]
@ethan-nelson many thanks for your useful comments. I will get back to you when I receive the second review and can more easily combine changes in the code in case of further feedback.

**Comment by mjsottile** [comment]
To update, I've reached out to the other reviewer via email to see what their status is.  Hopefully will have an update relatively soon.

**Comment by mjsottile** [comment]
I still have not heard back, so I am going to try to find an additional reviewer to un-stall the review.  Sorry for the delays @dmey.

**Comment by dmey** [comment]
> I still have not heard back, so I am going to try to find an additional reviewer to un-stall the review. Sorry for the delays @dmey.

Thank you @mjsottile and no problems!

**Comment by mjsottile** [comment]
:wave: @dvalters @AnsleyManke @andreas-h Would any of you be available/interested in providing a review for this JOSS submission?  I am currently looking for a reviewer to replace someone who has let me know that they are unable to complete the review.

**Comment by mjsottile** [comment]
:wave: @rchg @irnaka @jayten Hello - we are looking for an additional reviewer for this JOSS submission to replace a reviewer who can no longer do a review.  I found your names on the potential reviewers spreadsheet based on your preferred languages and domains of expertise.  Would any of you be available for a review?

**Comment by andreas-h** [comment]
Sorry, cannot do it.

**Comment by AnsleyManke** [comment]
This is not in my area of expertise, sorry.

**Comment by irnaka** [comment]
Sorry, this is not my area of expertise.

==============================
Theodosius Marwan IRNAKA

*marwan.irnaka@gmail.com <marwan.irnaka@gmail.com>*
*+62852 2803 5489 <%2B6285228035489>*
+336 47 08 81 33
* <marwan.irnaka@mail.ugm.ac.id>*

**Comment by jayten** [comment]
@mjsottile 
If you are still looking for a reviewer, I will be happy to help with this JOSS submission.

**Comment by mjsottile** [comment]
@whedon remove @rasmussn as reviewer

**Comment by whedon** [comment]
OK, @rasmussn is no longer a reviewer

**Comment by mjsottile** [comment]
@whedon add @jayten as reviewer

**Comment by whedon** [comment]
OK, @jayten is now a reviewer

**Comment by mjsottile** [comment]
Thank you @jayten for agreeing to review.  I’ve added you as a reviewer and updated the checklist at the top of the issue to include you.  Please let me know if you have any questions or need any assistance in completing your review.

**Comment by jayten** [comment]
I have gone through the review process for the [`teb`](https://github.com/teb-model/teb) repository.  I was able to install (on Ubuntu18 and MacOs 10.13) and run test cases with instructions provided in the documentation. The software solves the intended problem well. The software is well written and documented. However, the flavor of the FORTRAN code style is old, owing to the use of old legacy code. For example, "impnone.fh" just contain an 'implicit none' statement. Although code style is not a major issue, it can be improved over time with the contribution of others. I would suggest @dmey to add, at least a simple,  **CONTRIBUTING.md** for interested individuals to contribute to [`teb`](https://github.com/teb-model/teb). 

# License
I have issues with the license file: [CeCILL-C-V1](https://github.com/teb-model/teb/blob/joss-paper/Licence_CeCILL-C_V1-en.txt). The license file mentioned is not a part of the [OSI](https://opensource.org/licenses/alphabetical). Additionally, some old source files (like '[modei_flxsurf3db.f](https://github.com/teb-model/teb/blob/joss-paper/src/teb/modi_flxsurf3bx.f)') mention different license in the header, which is also not part of the [OSI](https://opensource.org/licenses/alphabetical). I want @mjsottile to take a decision on this issue. 

# Paper
@dmey , make sure all the contributors are at least acknowledged in the paper.md. For example, **P. Le Moigne**, mentioned as the author of the [*abort1_sfx.f90*](https://github.com/teb-model/teb/blob/joss-paper/src/driver/abor1_sfx.F90) source file, is neither mentioned in the author list nor the acknowledgment.
## References
Missing DOI for bib entry: rozoff_2003_J.Appl.Meteorol

**Comment by mjsottile** [comment]
@jayten Thanks for the great and prompt review.  I will look into the issues you raise when I return from travel next week.  I need to look more into the CeCILL licenses - as I understand them from looking at them a number of years ago, they are generally recognized as similar to the FSF and OSI licenses, but originated in Europe independently of the more common licenses that largely originated in the US.  I’ll do some reading and possibly ask the other JOSS editors for their opinions on the issue.

In the meantime @dmey, if you would like to look at the issues related to contributors and reference DOIs, that would be useful.

**Comment by danielskatz** [comment]
JOSS requires submitted software to have an OSI-approved license.  It looks like we (actually I) missed this originally, for which I apologize to the author and reviewers.

**Comment by danielskatz** [comment]
👋 @dmey - Is it possible to change (or dual license) with an OSI-approved license?  If not, we will have to ask you to withdraw this submission or we will have to reject it.

**Comment by danielskatz** [comment]
Note that https://opensource.org/licenses/CECILL-2.1 is an OSI-approved license - perhaps CeCILL-C-V1 is as well, but I can't find it on the OSI site.

**Comment by danielskatz** [comment]
Finally, I see on https://en.wikipedia.org/wiki/CeCILL: 

> Version 2.1 was released in June 2013. It allows relicensing to the GNU Affero General Public License and the European Union Public License as well as the GPL, and clarifies the language that requires licensees to give access to the source code (which had previously caused rejection of version 2.0 by the Open Source Initiative[3]).

> Note that CeCILL v1 already allowed replacing a CeCILL v1 license by CeCILL v2, so all software previously licensed with CeCILL v1 in 2004 can be licensed with CeCILL v2, with legal terms enforceable as authentic not only in French but in English too.

**Comment by dmey** [comment]
Hello all, many thanks for your prompt comments and feedback -- I will get back in the next 10 days or so.

**Comment by dmey** [comment]
@mjsottile and @danielskatz I have discussed these points with Masson given that he is the co-author who developed the physical model in 2000 and wrote most of the original code. We should be able to make most of the changes as requested but would to clarify a couple of points with you first before going ahead with the changes:
 
1.	Licence: OK -- this can be upgraded to CECILL-2.1 but will ask Masson to approve this as it will affect the whole software and not only my changes/enhancements.
2.	Acknowledgments: in the repository there are a couple of files authored by Le Moigne. I have double checked this again with Masson but it looks like these files were included from a 'sister' model (SURFEX) and Le Moigne was not a TEB contributor. Do you still require this to be acknowledged in the summary paper?
3.	Zenodo: We had not yet deposited the software on Zenodo however, given that my contribution is recent and specifically about the enhancements as described in the paper, I think that the software version reference should reflect all the software contributors including those acknowledged in the summary paper and with a different order – i.e. Masson et al. (2020).

**Comment by danielskatz** [comment]
The following are my thoughts as the AEiC on duty when this came up.  @mjsottile and the currently on-duty @openjournals/jose-eics might also have thoughts.

> 1. Licence: OK -- this can be upgraded to CECILL-2.1 but will ask Masson to approve this as it will affect the whole software and not only my changes/enhancements.

A license change needs to be approved by all contributors in theory, since they are the copyright holders.

> 2. Acknowledgments: in the repository there are a couple of files authored by Le Moigne. I have double checked this again with Masson but it looks like these files were included from a 'sister' model (SURFEX) and Le Moigne was not a TEB contributor. Do you still require this to be acknowledged in the summary paper?

If this software was not developed for the project but is included (in some sense, as a dependency), it likely makes sense to acknowledge that dependency. The author of this dependent software would not need to be an author of this software, in my opinion.

> 3. Zenodo: We had not yet deposited the software on Zenodo however, given that my contribution is recent and specifically about the enhancements as described in the paper, I think that the software version reference should reflect all the software contributors including those acknowledged in the summary paper and with a different order – i.e. Masson et al. (2020).

I'm not sure I understand this.  Both the JOSS paper and associated Zenodo (or other) archival repository deposit should contain the same list of authors, in the other determined by the authors, and with the list being determined by the authors.  The authors should include all significant software contributors, plus potentially others who contributed to the project in ways that may not be captured in the repository (e.g. software design, user support, ...)

**Comment by dmey** [comment]
> I'm not sure I understand this. Both the JOSS paper and associated Zenodo (or other) archival repository deposit should contain the same list of authors, in the other determined by the authors, and with the list being determined by the authors. The authors should include all significant software contributors, plus potentially others who contributed to the project in ways that may not be captured in the repository (e.g. software design, user support, ...)

I appreciate that this may be a bit confusing -- the reason I have asked is because we would rather not fork the original model but instead, improve and move the model to GitHub (i.e. this paper) and continue future development from this point forward. Therefore, whereas in the paper the authorship is clear for the work that was done in improving TEB, I think that for the associated Zenodo archive, it would be proper to keep the software name and authorship to reflect the contributions made over the years of development (i.e. leave name as TEB -- as also reflected in GitHub repository-- thus avoiding the creation of a separate fork for this specific submission). This is something that has been discussed with the TEB developers and seemed to make most sense.

**Comment by danielskatz** [comment]
👋 @arfon - any thoughts on the above comment?

**Comment by arfon** [comment]
If I understand the situation, the paper will have a shorter author list than the Zenodo archive? I don't think this is ideal but I think it's probably OK to allow this given the justification.

**Comment by mjsottile** [comment]
I agree.  It seems that the only outstanding item is to make sure all necessary parties consent to the update from CeCILL1 to CeCILL-2.1 such that the license will conform to the JOSS requirements.  Once you've indicated that is complete, I can scan through the state of the submission to push it to the next stage if it looks ready.

**Comment by mjsottile** [comment]
@whedon generate pdf from branch joss-paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss-paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02008/joss.02008/10.21105.joss.02008.pdf)

**Comment by mjsottile** [comment]
@whedon check references from branch joss-paper

**Comment by whedon** [comment]
```
Attempting to check references... from custom branch joss-paper
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.1023/A:1002463829265 is OK
- 10.1175/1520-0450(2004)043<0312:MTSEBO>2.0.CO;2 is OK
- 10.5194/gmd-5-1377-2012 is OK
- 10.5194/gmd-11-1929-2018 is OK
- 10.1002/joc.3415 is OK
- 10.5194/gmd-5-433-2012 is OK
- 10.5194/gmd-10-2801-2017 is OK
- 10.1175/WAF-D-11-00064.1 is OK
- 10.3389/fenvs.2014.00014 is OK
- 10.21105/joss.01137 is OK
- 10.5065/1dfh-6p97 is OK
- 10.1016/j.aeaoa.2019.100042 is OK
- 10.1007/s00703-008-0320-9 is OK
- 10.21105/joss.01468 is OK
- 10.1007/BF01025401 is OK
- 10.1007/s007030070003 is OK
- 10.1175/2009MWR2750.1 is OK
- 10.1007/s10546-009-9414-2 is OK
- 10.1175/1520-0493(1998)126<1373:TOCMGE>2.0.CO;2 is OK
- 10.1023/A:1016509614936 is OK
- 10.1007/s00585-997-0090-6 is OK
- 10.1007/s10546-006-9091-3 is OK
- 10.1175/2009JAMC2153.1 is OK
- 10.5281/zenodo.3562310 is OK

MISSING DOIs

- None

INVALID DOIs

- 10.1175/1520-0450(2002)041<1011%3AEOTTEB>2.0.CO%3B2 is INVALID
```

**Comment by mjsottile** [comment]
@dmey In addition to the license issue, both reviewers have not checked off the box related to community guidelines.  It looks like the paper provides guidance for users wishing to contribute, but the README or other documents in the repository do not.  You should add some text to the README or a separate CONTRIBUTING.md file for users who encounter the package but do not read the JOSS paper.

Also, it looks like there is an invalid DOI in the paper - can you repair that?  (See the "check references" result above).

Once you've had a chance to address these issues, please take a look at the reviewer comments again to make sure you've addressed both of their concerns.  At that point I'll check in with them again to see if they're satisfied with the outstanding boxes from their checklists.

**Comment by arfon** [comment]
Dear authors and reviewers

We wanted to notify you that in light of the current COVID-19 pandemic, JOSS has decided to suspend submission of new manuscripts and to handle existing manuscripts (such as this one) on a "best efforts basis". We understand that you may need to attend to more pressing issues than completing a review or updating a repository in response to a review. If this is the case, a quick note indicating that you need to put a "pause" on your involvement with a review would be appreciated but is not required.

Thanks in advance for your understanding.

_Arfon Smith, Editor in Chief, on behalf of the JOSS editorial team._

**Comment by dmey** [comment]
Many thanks all for your comments and useful feedback. Below I respond to each of your points individually.


Reviewer: @ethan-nelson

> ## General checks
> * License: The OSI list only indicates that CECILL-2.1 is open source compatible. It looks like there are some differences between CECILL-1, and I'm not sure if the OSI designation is backwards-compatible, so I will defer to @mjsottile for that.

1.1: This has been changed to [CECILL-2.1](https://github.com/teb-model/teb/blob/master/Licence_CeCILL_V2.1-en.txt).

> 
> ## Functionality
> * Installation: (CentOS 7.7, cmake 3.12.2, gcc 4.8.5) I ran into issues during compilation:
> 
> ```
> /home/enelson/review/teb/external/minimal-dx/src/shared/minimal_dx_fan.f90:99.64:
> 
>       error stop 'Fan Mode must either be 1 for on or 0 for off'
>                                                                 1
> Error: ERROR STOP statement not allowed in PURE procedure at (1)
> gmake[2]: *** [CMakeFiles/teb.dir/external/minimal-dx/src/shared/minimal_dx_fan.f90.o] Error 1
> gmake[1]: *** [CMakeFiles/teb.dir/all] Error 2
> gmake: *** [all] Error 2
> ```
> 
> After removing the error message, the model compiled just fine. This is a fairly old gcc version though so it may not warrant a change.

1.2: This issue has been addressed directly in the external library. Can you please let me know if TEB now compiles OK on your system or open an issue?

> ## Documentation
> * Automated tests: this is a corner case, but it is useful if the user could specify a custom cmake/gcc/git binary location for tests, either in the python call or near the top of one of the python scripts. My cmake is not in my PATH variable so I had to manually change the location in the python code. Adding this would allow for faster iteration over different versions of gcc, etc. for testing compatibility.

1.3: The current test scripts are meant to be used for running regression tests and not as as general framework for running regular TEB cases. With regards to your specific problem, did you try adding CMake to your current shell before running CMake if you do not want to add it permanently to your PATH? 

> * Community guidelines: Might I suggest a CONTRIBUTING.md file or indication in the README to complement the journal text? That could help direct folks who stumble upon the repo first as opposed to the paper.

1.4: `CONTRIBUTING.md` has been added.


> ## Other feedback:
> * A file in `docs/` with brief documentation detailing the necessary `input/` forcing files would be useful for tinkering from scratch. I ended up copying files from the example to play around with, but it would be good to have a list all the file variables and units--similar to the namelist docs.

1.5: This information has been added. Please see [input](https://github.com/teb-model/teb/blob/master/docs/software-docs.md#input).

> * Something else to consider: what are your thoughts on writing out a simple metadata file in the `output/` directory of a given run? Perhaps this onus should fall on the user to document their runs, but I could see potentially pulling in some metadata from the model build time (e.g. `git rev-parse`) that you could copy to `output/META.txt` or the like.

1.6: A TEB case directory comprises of a namelist, and input and output files. To retrieve metadata from the namelist we currently use the `get_date_params` function (see `teb/examples/CAPITOUL/tutorial.ipynb`). I am not yet sure what other metadata we should be saving but we could start discussing this in an issue for a future release.


Reviewer: @jayten

> I have gone through the review process for the [`teb`](https://github.com/teb-model/teb) repository. I was able to install (on Ubuntu18 and MacOs 10.13) and run test cases with instructions provided in the documentation. The software solves the intended problem well. The software is well written and documented. However, the flavor of the FORTRAN code style is old, owing to the use of old legacy code. For example, "impnone.fh" just contain an 'implicit none' statement. Although code style is not a major issue, it can be improved over time with the contribution of others. I would suggest @dmey to add, at least a simple, **CONTRIBUTING.md** for interested individuals to contribute to [`teb`](https://github.com/teb-model/teb).

2.1: Please see response to @ethan-nelson #1.4.

> 
> # License
> I have issues with the license file: [CeCILL-C-V1](https://github.com/teb-model/teb/blob/joss-paper/Licence_CeCILL-C_V1-en.txt). The license file mentioned is not a part of the [OSI](https://opensource.org/licenses/alphabetical). Additionally, some old source files (like '[modei_flxsurf3db.f](https://github.com/teb-model/teb/blob/joss-paper/src/teb/modi_flxsurf3bx.f)') mention different license in the header, which is also not part of the [OSI](https://opensource.org/licenses/alphabetical). I want @mjsottile to take a decision on this issue.

2.2: Please see response to @ethan-nelson #1.1.

> # Paper
> @dmey , make sure all the contributors are at least acknowledged in the paper.md. For example, **P. Le Moigne**, mentioned as the author of the [_abort1_sfx.f90_](https://github.com/teb-model/teb/blob/joss-paper/src/driver/abor1_sfx.F90) source file, is neither mentioned in the author list nor the acknowledgment.

2.3: I have double-checked this with the other authors and the list of TEB contributors under the acknowledgments is reported correctly. 

> ## References
> Missing DOI for bib entry: rozoff_2003_J.Appl.Meteorol

2.4: This has been added.


Editor/AEiC: @mjsottile/@danielskatz

All points should have been addressed. I only need to update one entry in the reference list (for which I am awaiting the DOI). With regards to Zenodo, I have added the full list of TEB contributors in the `zenodo.json` file. With regards to the licence, please see my response to @ethan-nelson #1.1.

**Comment by mjsottile** [comment]
Excellent!  It looks like everything has been addressed, but I do want to check on two items:

1. Can @ethan-nelson verify that 1.2 is addressed, and if not, open an issue as suggested by @dmey?

2. The license looks like it is addressed, but there is the caveat that @jayten pointed out that some of the files are from a third-party and have their own license wording in the comments (the "Environment Canada - Atmospheric Science and Technology License/Disclaimer").  The wording looks pretty permissive, but I am not a licensing expert.  Does @danielskatz have any input on this given that he's managed many more JOSS submissions than I have?

Thanks for your hard work on this @dmey.

**Comment by ethan-nelson** [comment]
1.2 is indeed fixed. Everything else looks good to me :+1:!

**Comment by mjsottile** [comment]
:wave: @arfon I haven't heard from @danielskatz so I assume he's busy at the moment.  Can you take a look at the issue I mention in my comment from last week?

**Comment by danielskatz** [comment]
sorry - I just missed the notification

**Comment by danielskatz** [comment]
> the caveat that @jayten pointed out that some of the files are from a third-party and have their own license wording in the comments (the "Environment Canada - Atmospheric Science and Technology License/Disclaimer"). The wording looks pretty permissive, but I am not a licensing expert. 

I'm afraid the answer is is that all software that is in the repo must have OSI-approved licenses.

**Comment by mjsottile** [comment]
@dmey I'm afraid the inclusion of files from a third-party that are under a different license than your project is a problem.  Looking at the list of files that appear to have the Environment Canada license text (with corresponding line counts):

```
matt:teb/ (master) $ wc -l `grep -lnr Canada src`
      23 src/teb/modi_init_surfconsphy.f
      71 src/teb/init_surfconsphy.F
      73 src/teb/stabfunc2.fh
      32 src/teb/vslog.f
      17 src/teb/impnone.cdk
      18 src/teb/surfcon.fh
      28 src/teb/consphy.cdk
      73 src/teb/stabfunc2.cdk
      33 src/teb/modi_flxsurf3bx.f
      18 src/teb/surfcon.cdk
      28 src/teb/consphy.fh
      17 src/teb/impnone.fh
     251 src/teb/flxsurf3bx.F
```

That is a relatively tiny proportion of the overall code.  Do you have any thoughts on how to proceed?

**Comment by danielskatz** [comment]
Just to be clear, the inclusion of files with a different license is not by itself the problem.

The problem is that 
1. the licenses for all files needs to be OSI-approved (listed in https://opensource.org/licenses/alphabetical) - the Environment Canada license does not appear to be
2. licenses need to be compatible, which is a more complicated discussion

**Comment by dmey** [comment]
> The license looks like it is addressed, but there is the caveat that @jayten pointed out that some of the files are from a third-party and have their own license wording in the comments (the "Environment Canada - Atmospheric Science and Technology License/Disclaimer"). The wording looks pretty permissive, but I am not a licensing expert. Does @danielskatz have any input on this given that he's managed many more JOSS submissions than I have?

@mjsottile @danielskatz @arfon. Environment Canada has kindly agreed to a change of licence for the files in question to MIT. I have now updated the repository to reflect this change.  Please could you let me know if you would like to see additional changes. Many thanks.

**Comment by mjsottile** [comment]
@dmey That looks like good progress.  The final thing you should do is verify on the CeCILL license documentation that you have the appropriate license for TEB.  Specifically, the documentation surrounding "borrowings" (e.g., the Environment Canada code) here:

https://cecill.info/placer.en.html

It appears that CeCILL-B is compatible with MIT.  It is not clear that the regular CeCILL 2.1 is.  Please see https://en.wikipedia.org/wiki/CeCILL#Other_CeCILL_licenses for a bit more info.  Once the we've verified that your project has the correct CeCILL licensing to be compatible with the borrowed MIT-licensed code, we can move to the next steps of finalizing the review.

**Comment by dmey** [comment]
@mjsottile Sorry but I am not sure what you mean exactly -- the Environment Canada (EC) code is licensed under MIT and bundled into TEB for redistribution (i.e. in a CeCILL 2.1 project). MIT is a (very) permissive licence with no copyleft. 

In [How to distribute my program under CeCILL page](https://cecill.info/placer.en.html) under [Inventory of the pieces of code and libraries that you use](https://cecill.info/placer.en.html#emprunt), the second paragraph reads:

> To be able to distribute your software under a given license, it is necessary that the licenses of these borrowings be compatible with the chosen license i.e. they should not conflict with license. In the case of CeCILL, the licenses that are compatible with the GNU GPL (and that can be found on [this page](https://www.gnu.org/licenses/license-list.html)) are compatible with CeCILL.

From the quote above, from [this page](https://www.gnu.org/licenses/license-list.html), the [Expat License](https://www.gnu.org/licenses/license-list.html#Expat) (i.e. MIT) is shown under the list of "GPL-Compatible Free Software Licenses".

**Comment by mjsottile** [comment]
@dmey Ok - I just wanted to make sure we had addressed @danielskatz second point re: license compatibility.  To my eye, it looks like we've addressed the license issue.  @danielskatz do you agree?  Assuming we're good I'll start my final pass over the review and materials.

**Comment by mjsottile** [comment]
@whedon generate pdf from branch joss-paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss-paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02008/joss.02008/10.21105.joss.02008.pdf)

**Comment by mjsottile** [comment]
@whedon check references from branch joss-paper

**Comment by whedon** [comment]
```
Attempting to check references... from custom branch joss-paper
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.5194/gmd-13-385-2020 is OK
- 10.5194/gmd-10-385-2017 is OK
- 10.1016/j.enbuild.2013.10.038 is OK
- 10.1023/A:1002463829265 is OK
- 10.1175/1520-0450(2004)043<0312:MTSEBO>2.0.CO;2 is OK
- 10.5194/gmd-5-1377-2012 is OK
- 10.5194/gmd-11-1929-2018 is OK
- 10.1002/joc.3415 is OK
- 10.5194/gmd-5-433-2012 is OK
- 10.5194/gmd-10-2801-2017 is OK
- 10.1175/WAF-D-11-00064.1 is OK
- 10.3389/fenvs.2014.00014 is OK
- 10.21105/joss.01137 is OK
- 10.5065/1dfh-6p97 is OK
- 10.1016/j.aeaoa.2019.100042 is OK
- 10.1007/s00703-008-0320-9 is OK
- 10.21105/joss.01468 is OK
- 10.1175/1520-0450(2003)042<0716:SOSLML>2.0.CO;2 is OK
- 10.1007/BF01025401 is OK
- 10.1007/s007030070003 is OK
- 10.1175/2009MWR2750.1 is OK
- 10.1007/s10546-009-9414-2 is OK
- 10.1175/1520-0493(1998)126<1373:TOCMGE>2.0.CO;2 is OK
- 10.1023/A:1016509614936 is OK
- 10.1007/s00585-997-0090-6 is OK
- 10.1007/s10546-006-9091-3 is OK
- 10.1175/2009JAMC2153.1 is OK
- 10.5281/zenodo.3562310 is OK
- 10.1175/1520-0450(2002)041<1011:EOTTEB>2.0.CO;2 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by mjsottile** [comment]
@dmey Looks good.  The only issue I see with the paper are some capitalizations in the references - the Rozoff title should have St. Louis, Missouri capitalized (it appears "st. Louis, missouri" in the PDF).  Similarly, the Xue paper title has bad capitalization.  I think in both cases it's just a matter of strategic placement of curly braces in the BibTeX title fields to keep the case correct in the generated output.

- Submit a version of the software to an archive (e.g., figshare or zenodo) and post the DOI for the archive in this thread.

- Please verify that the version number matches that of the archived code, and if not, let me know so I can update the version here.

- Finally, make sure that the metadata for the archive (e.g., title, authors) matches the paper.

Once all of these steps are done, I'll pass the paper on up the line to the EIC team for finalizing the process.

**Comment by danielskatz** [comment]
Re licenses, can @dmey please explain (maybe one more time, sorry), which licenses are currently used in the repository?

**Comment by dmey** [comment]
@whedon generate pdf from branch joss-paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss-paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02008/joss.02008/10.21105.joss.02008.pdf)

**Comment by dmey** [comment]
> @dmey Looks good. The only issue I see with the paper are some capitalizations in the references - the Rozoff title should have St. Louis, Missouri capitalized (it appears "st. Louis, missouri" in the PDF). Similarly, the Xue paper title has bad capitalization. I think in both cases it's just a matter of strategic placement of curly braces in the BibTeX title fields to keep the case correct in the generated output.

@mjsottile These have been fixed -- thank you. I am having an issue with the placement of Table 1 when this is typeset using the JOSS template. Would you know how to move Table 1 to page 2 so that the table is kept whole? I tried with a new line (using the HTML `<br>` tag) but that did not seem to have helped...

> Re licenses, can @dmey please explain (maybe one more time, sorry), which licenses are currently used in the repository?

Sure, no problem. To recap, the TEB project repository is licensed under CeCILL 2.1. The two 'external' codes that are currently bundled  (and redistributed) in TEB are: (a) subset of subroutines from the Mesoscale Compressible Community model from Environmental Canada (licensed under MIT) and (b) PsychroLib (licensed under MIT). The MinimalDX code is not bundled/redistributed with TEB as it is built and linked using CMake ExternalProject (as it usually done with other libraries components).

The two licences for the bundled code are in the `ThirdPartyNotices.txt` file.

**Comment by danielskatz** [comment]
Thanks - it seems that the license issue is resolved.

**Comment by mjsottile** [comment]
@dmey I didn't see any specific instructions in the JOSS docs but it does appear that you can use LaTeX commands to control formatting in pandoc (which is what whedon uses to generate the PDF).  See: https://stackoverflow.com/questions/16965490/pandoc-markdown-page-break

@arfon can weigh in - I think he knows Whedon best and would know how to do that.  Alternatively, you could break the paragraphs above the table up to force some whitespace to be inserted between paragraphs to force the table onto the next page and avoid inserting LaTeX commands into the markdown.

Once you're happy with the format of the paper, please see the points from my comment above (https://github.com/openjournals/joss-reviews/issues/2008#issuecomment-617407446) regarding archiving and we can finish the process.

**Comment by arfon** [comment]
> @arfon can weigh in - I think he knows Whedon best and would know how to do that. Alternatively, you could break the paragraphs above the table up to force some whitespace to be inserted between paragraphs to force the table onto the next page and avoid inserting LaTeX commands into the markdown.

Yeah, I think you can issue LaTeX commands such as `\newpage` to control the document flow.

**Comment by dmey** [comment]
@whedon generate pdf from branch joss-paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss-paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02008/joss.02008/10.21105.joss.02008.pdf)

**Comment by dmey** [comment]
Forcing page 1 to end with `\newpage` has worked and Table 1 is now rendered whole -- thanks.

> Once you're happy with the format of the paper, please see the points from my comment above ([#2008 (comment)](https://github.com/openjournals/joss-reviews/issues/2008#issuecomment-617407446)) regarding archiving and we can finish the process.

@mjsottile With regards to your last point in [#2008 (comment)](https://github.com/openjournals/joss-reviews/issues/2008#issuecomment-617407446), I added a [Zenodo metadata file](https://github.com/TEB-model/teb/blob/joss-paper/zenodo.json) to the repository. Can you please check that this is fine -- as we previously discussed, for Zenodo we wanted to keep consistent with previous and future version to avoid forking.

I am also waiting to hear back about a reference and DOI to be finalized so that I can update the list of references -- hopefully this will be done in the next couple of weeks. In the meantime, please let me know if there are other changes you would like to see.

**Comment by mjsottile** [comment]
:wave: @dmey - Just checking to see if you've heard back with the information you were waiting on to see if we're ready to move forward.  I believe that the metadata issue is fine with respect to keeping the repository sane and avoid unnecessary forking.

**Comment by dmey** [comment]
@mjsottile thanks for letting me know about the metadata. About reference and DOI, I have checked again but I was told that this will a little longer. I will let you know as soon as I hear back -- hopefully in the next few days.

**Comment by dmey** [comment]
@mjsottile just to let you know that I have enquired again about the status of this but was told that this will take another couple of weeks.

**Comment by kyleniemeyer** [comment]
Hi @dmey, just checking in on the status of this.

**Comment by dmey** [comment]
Perfect timing @kyleniemeyer -- I have now received the DOI. Thank you all for your patience. 
I have update the reference accordingly -- I will not check the final proof and update to Zenodo.

**Comment by dmey** [comment]
@whedon generate pdf from branch joss-paper

**Comment by whedon** [comment]
```
Attempting PDF compilation from custom branch joss-paper. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02008/joss.02008/10.21105.joss.02008.pdf)

**Comment by mjsottile** [comment]
Excellent.  Once you've posted the archive DOI and checked that the version number of the archive matches the version listed at the top of this thread, I'll run the final commands to hand off to the EIC team for the last steps.

**Comment by dmey** [comment]
@mjsottile, thank you. I have created 4.0.0 release and deposited at https://doi.org/10.5281/zenodo.3887081. Please let me know if you need more information.

**Comment by dmey** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.02008/joss.02008/10.21105.joss.02008.pdf)

**Comment by mjsottile** [comment]
@whedon set https://doi.org/10.5281/zenodo.3887081 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.3887081" target="_blank">10.5281/zenodo.3887081</a> is the archive.

**Comment by mjsottile** [comment]
@whedon accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary:

OK DOIs

- 10.5194/gmd-13-385-2020 is OK
- 10.5194/gmd-10-385-2017 is OK
- 10.1016/j.enbuild.2013.10.038 is OK
- 10.1023/A:1002463829265 is OK
- 10.1175/1520-0450(2004)043<0312:MTSEBO>2.0.CO;2 is OK
- 10.5194/gmd-5-1377-2012 is OK
- 10.5194/gmd-11-1929-2018 is OK
- 10.1002/joc.3415 is OK
- 10.5194/gmd-5-433-2012 is OK
- 10.5194/gmd-10-2801-2017 is OK
- 10.1175/WAF-D-11-00064.1 is OK
- 10.3389/fenvs.2014.00014 is OK
- 10.21105/joss.01137 is OK
- 10.5065/1dfh-6p97 is OK
- 10.1016/j.aeaoa.2019.100042 is OK
- 10.1007/s00703-008-0320-9 is OK
- 10.21105/joss.01468 is OK
- 10.1175/1520-0450(2003)042<0716:SOSLML>2.0.CO;2 is OK
- 10.1007/BF01025401 is OK
- 10.1007/s007030070003 is OK
- 10.1175/2009MWR2750.1 is OK
- 10.1007/s10546-009-9414-2 is OK
- 10.1175/1520-0493(1998)126<1373:TOCMGE>2.0.CO;2 is OK
- 10.1023/A:1016509614936 is OK
- 10.1007/s00585-997-0090-6 is OK
- 10.1007/s10546-006-9091-3 is OK
- 10.1175/2009JAMC2153.1 is OK
- 10.5281/zenodo.3562310 is OK
- 10.1175/1520-0450(2002)041<1011:EOTTEB>2.0.CO;2 is OK

MISSING DOIs

- None

INVALID DOIs

- 10.1029/2019MS001961 is INVALID
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/1483

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/1483, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by kyleniemeyer** [comment]
Hi @dmey - can you confirm the DOI 10.1029/2019MS001961 for your 2020 JAMES article? Is that an upcoming article that is not online yet?

**Comment by dmey** [comment]
Hi @kyleniemeyer. That is correct -- the DOI was sent out to me today but it may take a few more days before it becomes active.

**Comment by kyleniemeyer** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1270485465267933185) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/1485
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.02008](https://doi.org/10.21105/joss.02008)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? notify your editorial technical team...

**Comment by kyleniemeyer** [comment]
Congratulations  @dmey on your article's publication in JOSS!
Many thanks to @ethan-nelson and @jayten for reviewing this submission, and @mjsottile for editing it.

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.02008/status.svg)](https://doi.org/10.21105/joss.02008)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.02008">
  <img src="https://joss.theoj.org/papers/10.21105/joss.02008/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.02008/status.svg
   :target: https://doi.org/10.21105/joss.02008
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.02008/status.svg)](https://doi.org/10.21105/joss.02008)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss

**Comment by dmey** [comment]
Thank you @kyleniemeyer. Many thanks again to editors and reviewers for their useful feedback.
